### ISSUE TITLE
Implementing "move part to shard" feature with Fixes

### ISSUE THREAD

**Issue by tarun-annapareddy** [ClickHouse/ClickHouse#65633]
### Issue

ClickHouse currently lacks a rebalancing feature, which forces users to perform manual rebalancing when shards become overloaded and need horizontal scaling by adding new shards. This manual process can lead to data inconsistencies during the execution of `FETCH`, `ATTACH`, and `DELETE` commands due to merges and mutations.

The proposal in [Issue 13574](https://github.com/ClickHouse/ClickHouse/issues/13574) and the PR [17871](https://github.com/ClickHouse/ClickHouse/pull/17871) introduced the `ALTER TABLE xxx MOVE PART 'xxx' TO SHARD 'xxx'` command. It is an excellent idea, to solve the problem. But, Its implementation faces some challenges:

1. **Use of partUUID**: Many users have not assigned UUIDs to parts, making their usage difficult. Additionally, the community has shown limited interest in introducing UUIDs, necessitating the exploration of alternative methods.
2. **Inefficient Task Checking and Resource wastage**: The orchestrator repeatedly checks if the log has been picked by the replica instead of using callbacks, making the process resource-intensive. The orchestrator waits for replicas to pull the task from `/logs` to their queue

### Proposal

To address these challenges, I propose a simpler and more reliable way to move parts between shards. The goal is to provide users with a reliable move-part solution without overcomplicating.

#### Fix-1: Using Part Hash and Handling Read/Updates

1. **Creation of Pinned Part Block IDs**:
   - Replace `pinned_part_uuids` with `pinned_part_block_ids`.
   - Generate these hashes using `part->checksums.computeTotalChecksumDataOnly()`, resulting in a 128-bit `block_id`. This `block_id` is computed once when needed and stored in `IMergeTreeDataPart` for direct access.
   - A sample `block_id` might look like `202402_12482096457872641557_2282438499454264908`, similar to block IDs used for write deduplication.

2. **Delaying Merges and Mutations**:
   - Modify the merge selection algorithm to exclude these parts from merge selection and mutation operations.

3. **Blocking Reads**:
   - Block reads on pinned parts by comparing the `block_id` of the selected parts with `pinned_part_block_ids`.
   - Throw a "Part under migration" exception from `ReplicatedMergeTree`.
   
   Also, the assumption is that we do not have duplicate parts data hashes. It feels a reasonable assumption given that Clickhouse already implements write de-duplication for a window. I want to receive community feedback for this.

#### Fix-2: Orchestrator with Callbacks

1. **ZooKeeper Node Structure**:
   - Maintain two Znodes inside `/zookeeper/part_moves_shard`:
     - `tasks/task_uuid`: Stores metadata and the task state.
     - `task_queue/<task_uuid>`: Stores signals for execution.

2. **Background Job Assignment**:
   - Create a `BackgroundJobsAssignee` similar to the DataProcessing job.
   - This job periodically wakes up and executes background move part tasks.

3. **Sample Task Structure**:
   - `{Entry_STATE, create_time, update_time, part_name, block_id, to_shard, status,required_number_of_replicas, replicas}`

4. **Move Part Command Execution**:
   - When the Move Part command is executed:
     - A task is created in `tasks/task_uuid` path with metadata `{TODO, create_time, update_time, part_name, block_id, to_shard}`.
     - `<task_uuid>` is added to the `task_queue`.

6. **Task Orchestration**:
   - `PartMovesBetweenShardsOrchestrator` periodically wakes up and picks a task from the queue.
   - It checks `tasks/task_uuid/replica` to see if the task is already picked for execution.
   - If not, the current replica adds itself as `replica` by creating an ephemeral node and deleting `task_uuid` from the queue.

7. **Task Execution**:
   - The orchestrator verifies if the current state is marked as successful.
   - If successful, it schedules the next state with the `required_number_of_replicas` (all replicas in the shard where the log is executed) and writes the necessary log to `/zookeeper/logs/log-`.
   - Each replica of a shard reads the log and executes the command.
   - Once the command is executed, they update the replicas.
   - The last replica that reaches the `required_number_of_replicas`, marks the task as successful and writes `task_uuid` into the task_queue.
   - If task execution fails, the task will be retried by the ClickHouse queue.

8. **Failed Task Handling**:
   - Along with executing new tasks, the orchestrator checks tasks picked by it and updated before a threshold period and marks them as FAILED.
   - Necessary compensation actions are performed.

I am tagging all the people who worked on the `Move PART TO SHARD` discussion. Please provide your insights on this. I am willing to spend the dev hours and submit PR. cc: @alesapin @alexey-milovidov @azat @nvartolomei @xjewer

**Comment by UnamedRus** [comment]
Use case described in this feature request: 

1. (Atomically) move part across 2 different ReplicatedMergeTree tables

Sometimes, there is demand for multilayer ClickHouse installations, and few additional options actually possible:

2. Move part across ReplicatedMergeTree table and simple MergeTree (for example backed by s3_plain disk)
3. Move data across ReplicatedMergeTree table and table engine, like S3/Azure/HDFS and others.

And when people are willing to implement their custom "Move logic", biggest problem is sync for reads in order to avoid double read of part or no read at all.

One of interesting options is to use:

> │ additional_table_filters │ {}    │       0 │ Additional filter expression which would be applied after reading from specified table. Syntax: {'table1': 'expression', 'database.table2': 'expression'} │ ᴺᵁᴸᴸ │ ᴺᵁᴸᴸ │        0 │ Map    │ {}      │           │   

or ROW POLICIES, for manipulation of parts read on each node.

No atomicity across nodes are actually required, because there is only one initiator node (Distributed table), which build that particular query for other nodes and which will use some state (in our example user configuration stored in ZK) for actually deciding which "shard" are actually going to read our moving part.

If we assume, that data itself is not changed during moving, it's not a problem if particular "late" initiator node will read moved part from source shard.

The main problem with that approach is missing ability to stop merges on part, because if part will be merged on source node, it does mean that our condition to skip this particular part will be missed.
Secondary problem: We do not control part name on destination for use case 1, (but can for 2 and 3)


Whole pipeline could look like

1. ALTER USER MODIFY SETTING additional_table_filters = {'db.table': 'indexHint(NOT (shardNum() = "destination" AND _part/_file = "XXXXXXXXX'') }
2. actual moving of part, can be done via FETCH PART or whatever option
3. ALTER USER MODIFY SETTING additional_table_filters = {'db.table': 'indexHint(NOT (shardNum() = "source" AND _part = "XXXXXXXXX'') }
4. After all nodes updated user configuration, drop part on source node.

**Comment by tarun-annapareddy** [comment]
Hi, @UnamedRus This is to fix bugs or re-implement the already written "MOVE PART TO SHARD". Let's stick to this command itself and not derail it by bringing multiple requirements. That's why I did not mark it as a new feature-request. All these discussions have already been done before in base proposal/PR.

**Comment by UnamedRus** [comment]
My comment, is to highlight one particular shared problem:

1. Ability to start/stop merges for particular part. (And some user interface for that)

As, better implementation of MOVE PART TO SHARD is interesting and great goal to have, but as any other complex task, complete production ready and bulletproof implementation can take long time if even succeed. 

Ability to stop merges, is much smaller task, which can be useful by itself even if main goal will not be achieved.

**Comment by tarun-annapareddy** [comment]
@UnamedRus I completely agree, the ability to stop merges at partition/part level will help users. I initially started working on proposing that feature and figured out that "MOVE PART" gives users much more leverage. We will discuss community interest in "Move PART" and work towards the "STOP MERGES ON PART" command if this is not working out

**Comment by tarun-annapareddy** [comment]
@alesapin @alexey-milovidov @azat @nvartolomei @xjewer @UnamedRus I have implemented the second phase improvement of "Move Parts" and raised PR. I wanted to have feedback from the community to make incremental improvements and make this feature production level.

**Comment by azat** [comment]
@tarun-annapareddy I can't promise that someone will take a look (since AFAIU this feature is questionable at least in the current implementation), but it is better to submit your changes to the official repo to get more attention

**Comment by tarunannapareddy** [comment]
Hi @azat Thanks for your reply. I have implemented the phase-2 of above idea and raised the PR 
 (https://github.com/ClickHouse/ClickHouse/pull/68020)
